src/integrationTest/java/com/hivemq/client/restrictions/Mqtt3SendMaximumIT.java

```diff
@@ -0,0 +1,123 @@
+import com.hivemq.extension.sdk.api.interceptor.puback.parameter.PubackOutboundOutput;
+import com.hivemq.extension.sdk.api.parameter.ExtensionStartInput;
+import com.hivemq.extension.sdk.api.parameter.ExtensionStartOutput;
+import com.hivemq.extension.sdk.api.parameter.ExtensionStopInput;
+import com.hivemq.extension.sdk.api.parameter.ExtensionStopOutput;
+import com.hivemq.extension.sdk.api.services.Services;
+import com.hivemq.extension.sdk.api.services.intializer.ClientInitializer;
+import com.hivemq.testcontainer.core.HiveMQExtension;
+import com.hivemq.testcontainer.junit5.HiveMQTestContainerExtension;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import java.time.Duration;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/**
+ * @author Yannick Weber
+ */
+public class Mqtt3SendMaximumIT {
+
+    public static final int RECEIVE_MAXIMUM = 10;
+    public static final @NotNull HiveMQExtension NO_PUBACK_EXTENSION = HiveMQExtension.builder()
+            .version("1.0.0")
+            .priority(100)
+            .name("No PUBACK Extension")
+            .id("no-puback-extension")
+            .mainClass(NoPubackExtension.class)
+            .build();
+
+    @RegisterExtension
+    public final @NotNull HiveMQTestContainerExtension hivemq =
+            new HiveMQTestContainerExtension().withExtension(NO_PUBACK_EXTENSION);
+
+    @Test
+    void test() throws InterruptedException {
+
+        final Mqtt3Client publisher = Mqtt3Client.builder().serverPort(hivemq.getMqttPort()).build();
+        publisher.toBlocking().connectWith().restrictions().sendMaximum(RECEIVE_MAXIMUM).applyRestrictions().send();
+
+        Mqtt3Connect.builder().restrictions().sendMaximum(10).sendMaximumPacketSize().applyRestrictions().build();
+
+        final ConcurrentLinkedQueue<Mqtt5Publish> publishes = new ConcurrentLinkedQueue<>();
+        final Mqtt5BlockingClient subscriber = Mqtt5Client.builder().serverPort(hivemq.getMqttPort()).buildBlocking();
+        subscriber.connectWith().send();
+        subscriber.toAsync().publishes(MqttGlobalPublishFilter.ALL, publishes::add);
+        subscriber.subscribeWith().topicFilter("#").send();
+
+        for (int i = 0; i < 12; i++) {
+            publisher.toAsync()
+                    .publishWith()
+                    .topic("test")
+                    .qos(MqttQos.AT_LEAST_ONCE)
+                    .send()
+                    .whenComplete((mqtt3Publish, throwable) -> {
+                        if (mqtt3Publish != null) {
+                            System.out.println("PUBACK");
+                        }
+                        if (throwable != null) {
+                            System.out.println("ERROR");
+                        }
+                    });
+        }
+
+        Thread.sleep(10000);
```

Comment:
that doesnt feel good, why not just waiting for RECEIVE_MAXIMUM = publishes.size?

Reply:
maybe some other assertion might be usefull here

Reply:
This needs to be done, because testing for "events are not happening" is not possible.